src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/shmem_port.h

```diff
@@ -5,83 +5,82 @@
 #ifndef SHMEM_PORT_H_
 #define SHMEM_PORT_H_
 
-#include <pybind11/numpy.h>
-
-#include <thread>  // NOLINT [build/c++11]
+#include <pthread.h>
 #include <queue>
 #include <string>
+#include <vector>
+#include <atomic>
+#include <mutex>  // NOLINT
+#include <condition_variable>  // NOLINT
 
 #include "abstract_port.h"
 #include "shm.h"
 #include "utils.h"
 
 namespace message_infrastructure {
 
-template<class T>
-class ShmemRecvQueue {
- public:
-  T get(bool block = true, time_t timeout = 0, bool peek = false);
- private:
-  std::queue<T> queue_;
-};
-template<class T>
 class ShmemSendPort : public AbstractSendPort {
  public:
   ShmemSendPort(const std::string &name,
-                SharedMemoryPtr shm,
-                Proto *proto,
+                SharedMemory shm,
                 const size_t &size,
-                sem_t *req,
-                sem_t *ack);
+                const size_t &nbytes);
   std::string Name();
-  pybind11::dtype Dtype();
-  ssize_t* Shape();
   size_t Size();
   int Start();
   int Probe();
-  int Send();
+  int Send(void* data);
   int Join();
   int AckCallback();
 
-  SharedMemoryPtr shm_ = NULL;
-  sem_t *req_ = NULL;
-  sem_t *ack_ = NULL;
+  SharedMemory shm_;
   int idx_ = 0;
-  bool done_ = false;
+  std::atomic_bool done_;
   void *array_ = NULL;
   sem_t *semaphore_ = NULL;
   void *observer = NULL;
-  std::thread *thread_ = NULL;
 };
-template<class T>
+
+class ShmemRecvQueue {
+ public:
+  void Init(const size_t &capacity, const size_t &nbytes);
+  void Push(void* src);
+  void Pop();
+  void* Front();
+  void* FrontPop();
+  bool Probe();
+
+ private:
+  std::mutex lock_;
+  std::condition_variable cond_;
+  size_t nbytes_;
+  size_t capacity_;
+  std::queue<void *> queue_;
+};
+
 class ShmemRecvPort : public AbstractRecvPort {
  public:
   ShmemRecvPort(const std::string &name,
-                SharedMemoryPtr shm,
-                Proto *proto,
+                SharedMemory shm,
                 const size_t &size,
-                sem_t *req,
-                sem_t *ack);
+                const size_t &nbytes);
   std::string Name();
-  pybind11::dtype Dtype();
-  ssize_t* Shape();
   size_t Size();
   int Start();
-  int Probe();
-  int Recv();
+  bool Probe();
+  void* Recv();
   int Join();
```

Comment:
do we need to add `Stop`?
- set done_ to True.

at `Join`, we need to invoke thread.join and free data at queue_.